ProcessMaker/processmaker
# Handling multiple sessions and tabs - FOUR-6985

## Issue & Reproduction Steps

The QA team has reported other issues related to the `AccountTimeout` feature.
- Multiple tabs in the same session
- Handling different sessions of the same user

## Solution
Create a device identifier. For simplicity, the identification is based on the session cookie. In order not to expose the session cookie in javascript, a secure hash was used. During testing I also noticed that the session cookie name was not being respected, as well as the `expire_on_close` setting.

Once the device identifier and cookies were properly configured, the next step was to work with 2 events: SessionStarted and Logout and ensure that the javascript only accepts events from the same identifier.

## How to Test

Put the following settings in the .env file.
```
SESSION_LIFETIME=1
SESSION_EXPIRE_WARNING=30
```
Log in and do not check `Remember me`.

- Wait for 30 seconds and click `Stay connected`
- Open 2 tabs in the same session and click 1x on `Stay connected`, 1x on `Logout` and 1x don't click on anything.
- Open 2 tabs in the same session and in one tab click on `Logout` menu below `About`.
- Log in 2 different browsers with the same user and click 1x on `Stay connected` in, 1x on `Logout` and 1x do not click on anything..

## Related Tickets & Packages
- [FOUR-6985](https://processmaker.atlassian.net/browse/FOUR-6985)

## Code Review Checklist
